### ISSUE TITLE
Consider using UTF8 literals for string literals generated from .razor and .cshtml

### ISSUE THREAD

**Issue by tmat** [dotnet/aspnetcore#50788]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

The source generator currently uses regular string literals for HTML content.

For example,
```C#
__builder.AddMarkupContent(4, "<h1>Hello, world!</h1>\n\nWelcome to your new app!\n\n");
```

String literals are stored in the assembly in `#US` metadata heap with UTF16. The heap size is limited and larger apps run out of the space.
See https://github.com/dotnet/aspnetcore/issues/20224.

### Describe the solution you'd like

Using UTF8 literals would address this issue. UTF8 literals are stored directly in the PE image at certain RVA. The metadata heap size limitations thus do not apply.

Besides, it would likely be faster and more memory efficient to render the pages since UTF16 to UTF8 encoding might be avoided.

**Comment by justindbaur** [comment]
Would a community contribution of this be accepted?

**Comment by DamianEdwards** [comment]
Note I started investigating a fix to support this a while back in https://github.com/dotnet/razor/tree/damianedwards/Utf8HtmlLiterals
Can't quite remember where I got to, but I think it involved a new directive to enable changing the format the HTML literals are encoded in and plumbing that all the way through to the compiler.

**Comment by jjonescz** [comment]
Implementing this in the razor compiler seems straightforward, but we would need APIs from the aspnetcore runtime which we could call with `ReadOnlySpan<byte>` (UTF8 strings) instead of `string` (UTF16 strings).

For example:
- For razor components, we need UTF8 overloads on the RenderTreeBuilder which presumably needs new RenderTreeFrame type for UTF8 string and its implementation in the actual renderers.
- Razor pages use `TextWriter` under the hood which itself only supports UTF16 strings, so I'm not sure how we would add UTF8 overloads there
  https://github.com/dotnet/aspnetcore/blob/fe0fbff040c9f129ba3f5ee8cf15c8f6d96fb50e/src/Mvc/Mvc.Razor/src/RazorPageBase.cs#L467

**Comment by DamianEdwards** [comment]
Yep the compiler changes are one side of this, the other is updating the view engines/frameworks to support it. We could decide to enable one or both of .cshtml and .razor based on priorities.

For .cshtml one approach would be to just add overloads of the applicable Write methods that accept a byte array on the various razor base classes and have them convert the UTF8 bytes to strings, but that's almost certainly going to be hard to do efficiently. Alternatively the MVC view engines rendering pipeline would likely need to be updated to either support interleaving string and UTF8 bytes values, or overhauled to be based on bytes instead of TextWriter.
